src/main/java/com/sumologic/log4j/aggregation/BufferFlusherThread.java

```diff
@@ -96,17 +104,44 @@ protected BufferFlushingTask(BufferWithEviction<In> messageQueue)
 
   /* Public interface */
 
+  public void setTerminating()
+  {
+    terminating = true;
+  }
+
+  public void runTask()
+  {
+    if ((terminating && messageQueue.size() > 0) || needsFlushing()) {
+      flushAndSendCatchingExceptions();
+    }
+  }
+
   @Override
   public void run()
   {
-    if (needsFlushing()) {
+    while (true) {
+      runTask();
+      if (terminating) {
```

Comment:
There's a race condition here. If it doesn't need flushing but terminate is called before this line, then you will loose data.

Can you change this to capture the value of terminating after the `while (true)` in a `final` local variable, and pass that to `runTask` as well as being used here?